File: android/lib/src/main/kotlin/com/noise/protocol/crypto/CipherSuite.kt
package com.noise.protocol.crypto

import blake.hash.BLAKE2b
import blake.hash.BLAKE2s
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec

/**
 * Bundles all cryptographic operations for a Noise cipher suite.
 */
class CipherSuite(
    val dhName: String,
    val cipherName: String,
    val hashName: String,
    val dhlen: Int,
    val hashlen: Int,
    val blocklen: Int,
    val generateKeyPair: () -> NoiseKeyPair,
    val dh: (NoiseKeyPair, ByteArray) -> ByteArray,
    val keyPairFromPrivate: (ByteArray) -> NoiseKeyPair,
    val encrypt: (ByteArray, Long, ByteArray, ByteArray) -> ByteArray,
    val decrypt: (ByteArray, Long, ByteArray, ByteArray) -> ByteArray,
    val hash: (ByteArray) -> ByteArray,
    val hmacHash: (ByteArray, ByteArray) -> ByteArray
) {
    /** HKDF derived from hmacHash. */
    fun hkdf(chainingKey: ByteArray, inputKeyMaterial: ByteArray, numOutputs: Int): List<ByteArray> {
        val tempKey = hmacHash(chainingKey, inputKeyMaterial)
        val output1 = hmacHash(tempKey, COUNTER_01)
        val input2 = ByteArray(output1.size + 1)
        System.arraycopy(output1, 0, input2, 0, output1.size)
        input2[output1.size] = 0x02
        val output2 = hmacHash(tempKey, input2)
        if (numOutputs == 2) return listOf(output1, output2)
        val input3 = ByteArray(output2.size + 1)
        System.arraycopy(output2, 0, input3, 0, output2.size)
        input3[output2.size] = 0x03
        val output3 = hmacHash(tempKey, input3)
        return listOf(output1, output2, output3)
    }

    /** Construct full protocol name. */
    fun protocolName(pattern: String): String =
        "Noise_${pattern}_${dhName}_${cipherName}_${hashName}"

    companion object {
        private val COUNTER_01 = byteArrayOf(0x01)

        // Shared DH lambdas for all X25519 suites
        private val x25519Generate: () -> NoiseKeyPair = { NoiseKeyPair.generate() }
        private val x25519DH: (NoiseKeyPair, ByteArray) -> ByteArray = { kp, pub -> kp.dh(pub) }
        private val x25519FromPrivate: (ByteArray) -> NoiseKeyPair = { NoiseKeyPair.fromPrivateKey(it) }

        // ChaChaPoly lambdas
        private val chachaEncrypt: (ByteArray, Long, ByteArray, ByteArray) -> ByteArray =
            { k, n, ad, pt -> NoiseCipher.encrypt(k, n, ad, pt) }
        private val chachaDecrypt: (ByteArray, Long, ByteArray, ByteArray) -> ByteArray =
            { k, n, ad, ct -> NoiseCipher.decrypt(k, n, ad, ct) }

        // AESGCM lambdas
        private val aesgcmEncrypt: (ByteArray, Long, ByteArray, ByteArray) -> ByteArray =
            { k, n, ad, pt -> NoiseCipherAESGCM.encrypt(k, n, ad, pt) }
        private val aesgcmDecrypt: (ByteArray, Long, ByteArray, ByteArray) -> ByteArray =
            { k, n, ad, ct -> NoiseCipherAESGCM.decrypt(k, n, ad, ct) }

        // SHA-256 lambdas
        private val sha256Hash: (ByteArray) -> ByteArray = { NoiseHash.hash(it) }
        private val sha256HmacHash: (ByteArray, ByteArray) -> ByteArray =
            { key, data -> NoiseHash.hmacHash(key, data) }

        // SHA-512 lambdas
        private val sha512Hash: (ByteArray) -> ByteArray = { NoiseHashSHA512.hash(it) }
        private val sha512HmacHash: (ByteArray, ByteArray) -> ByteArray =
            { key, data -> NoiseHashSHA512.hmacHash(key, data) }

        // BLAKE2s lambdas — streaming HMAC avoids ByteArray concatenation
        private val blake2sHash: (ByteArray) -> ByteArray = { BLAKE2s.hash(it) }
        private val blake2sHmacHash: (ByteArray, ByteArray) -> ByteArray =
            { key, data -> hmacBlake2s(key, data) }

        // BLAKE2b lambdas — streaming HMAC avoids ByteArray concatenation
        private val blake2bHash: (ByteArray) -> ByteArray = { BLAKE2b.hash(it) }
        private val blake2bHmacHash: (ByteArray, ByteArray) -> ByteArray =
            { key, data -> hmacBlake2b(key, data) }

        /** Streaming HMAC-BLAKE2s (RFC 2104) — avoids ipad+data concatenation. */
        private fun hmacBlake2s(key: ByteArray, data: ByteArray): ByteArray {
            var k = if (key.size > 64) BLAKE2s.hash(key) else key
            if (k.size < 64) k = k.copyOf(64)
            val pad = ByteArray(64)
            // Inner: BLAKE2s(ipad || data)
            for (i in 0 until 64) pad[i] = (k[i].toInt() xor 0x36).toByte()
            val innerHash = BLAKE2s.Hasher().update(pad).update(data).finalize()
            // Outer: BLAKE2s(opad || innerHash)
            for (i in 0 until 64) pad[i] = (k[i].toInt() xor 0x5c).toByte()
            return BLAKE2s.Hasher().update(pad).update(innerHash).finalize()
        }

        /** Streaming HMAC-BLAKE2b (RFC 2104) — avoids ipad+data concatenation. */
        private fun hmacBlake2b(key: ByteArray, data: ByteArray): ByteArray {
            var k = if (key.size > 128) BLAKE2b.hash(key) else key
            if (k.size < 128) k = k.copyOf(128)
            val pad = ByteArray(128)
            // Inner: BLAKE2b(ipad || data)
            for (i in 0 until 128) pad[i] = (k[i].toInt() xor 0x36).toByte()
            val innerHash = BLAKE2b.Hasher().update(pad).update(data).finalize()
            // Outer: BLAKE2b(opad || innerHash)
            for (i in 0 until 128) pad[i] = (k[i].toInt() xor 0x5c).toByte()
            return BLAKE2b.Hasher().update(pad).update(innerHash).finalize()
        }

        // --- The 8 cipher suites ---

        val NOISE_25519_CHACHAPOLY_SHA256 = CipherSuite(
            dhName = "25519", cipherName = "ChaChaPoly", hashName = "SHA256",
            dhlen = DHLEN, hashlen = 32, blocklen = 64,
            generateKeyPair = x25519Generate, dh = x25519DH, keyPairFromPrivate = x25519FromPrivate,
            encrypt = chachaEncrypt, decrypt = chachaDecrypt,
            hash = sha256Hash, hmacHash = sha256HmacHash
        )

        val NOISE_25519_CHACHAPOLY_SHA512 = CipherSuite(
            dhName = "25519", cipherName = "ChaChaPoly", hashName = "SHA512",
            dhlen = DHLEN, hashlen = 64, blocklen = 128,
            generateKeyPair = x25519Generate, dh = x25519DH, keyPairFromPrivate = x25519FromPrivate,
            encrypt = chachaEncrypt, decrypt = chachaDecrypt,
            hash = sha512Hash, hmacHash = sha512HmacHash
        )

        val NOISE_25519_CHACHAPOLY_BLAKE2S = CipherSuite(
            dhName = "25519", cipherName = "ChaChaPoly", hashName = "BLAKE2s",
            dhlen = DHLEN, hashlen = 32, blocklen = 64,
            generateKeyPair = x25519Generate, dh = x25519DH, keyPairFromPrivate = x25519FromPrivate,
            encrypt = chachaEncrypt, decrypt = chachaDecrypt,
            hash = blake2sHash, hmacHash = blake2sHmacHash
        )

        val NOISE_25519_CHACHAPOLY_BLAKE2B = CipherSuite(
            dhName = "25519", cipherName = "ChaChaPoly", hashName = "BLAKE2b",
            dhlen = DHLEN, hashlen = 64, blocklen = 128,
            generateKeyPair = x25519Generate, dh = x25519DH, keyPairFromPrivate = x25519FromPrivate,
            encrypt = chachaEncrypt, decrypt = chachaDecrypt,
            hash = blake2bHash, hmacHash = blake2bHmacHash
        )

        val NOISE_25519_AESGCM_SHA256 = CipherSuite(
            dhName = "25519", cipherName = "AESGCM", hashName = "SHA256",
            dhlen = DHLEN, hashlen = 32, blocklen = 64,
            generateKeyPair = x25519Generate, dh = x25519DH, keyPairFromPrivate = x25519FromPrivate,
            encrypt = aesgcmEncrypt, decrypt = aesgcmDecrypt,
            hash = sha256Hash, hmacHash = sha256HmacHash
        )

        val NOISE_25519_AESGCM_SHA512 = CipherSuite(
            dhName = "25519", cipherName = "AESGCM", hashName = "SHA512",
            dhlen = DHLEN, hashlen = 64, blocklen = 128,
            generateKeyPair = x25519Generate, dh = x25519DH, keyPairFromPrivate = x25519FromPrivate,
            encrypt = aesgcmEncrypt, decrypt = aesgcmDecrypt,
            hash = sha512Hash, hmacHash = sha512HmacHash
        )

        val NOISE_25519_AESGCM_BLAKE2S = CipherSuite(
            dhName = "25519", cipherName = "AESGCM", hashName = "BLAKE2s",
            dhlen = DHLEN, hashlen = 32, blocklen = 64,
            generateKeyPair = x25519Generate, dh = x25519DH, keyPairFromPrivate = x25519FromPrivate,
            encrypt = aesgcmEncrypt, decrypt = aesgcmDecrypt,
            hash = blake2sHash, hmacHash = blake2sHmacHash
        )

        val NOISE_25519_AESGCM_BLAKE2B = CipherSuite(
            dhName = "25519", cipherName = "AESGCM", hashName = "BLAKE2b",
            dhlen = DHLEN, hashlen = 64, blocklen = 128,
            generateKeyPair = x25519Generate, dh = x25519DH, keyPairFromPrivate = x25519FromPrivate,
            encrypt = aesgcmEncrypt, decrypt = aesgcmDecrypt,
            hash = blake2bHash, hmacHash = blake2bHmacHash
        )
    }
}

File: ios/Sources/NoiseProtocol/Crypto/CipherSuite.swift
import Foundation
import BlakeHash
import CBLAKE2

/// Bundles all cryptographic operations for a Noise cipher suite.
public struct CipherSuite: Sendable {
    public let dhName: String
    public let cipherName: String
    public let hashName: String
    public let dhlen: Int
    public let hashlen: Int
    public let blocklen: Int

    // DH operations
    public let generateKeyPair: @Sendable () -> NoiseKeyPair
    public let dh: @Sendable (NoiseKeyPair, Data) throws -> Data
    public let keyPairFromPrivate: @Sendable (Data) throws -> NoiseKeyPair

    // Cipher operations (key is raw Data, 32 bytes)
    public let encrypt: @Sendable (Data, UInt64, Data, Data) throws -> Data
    public let decrypt: @Sendable (Data, UInt64, Data, Data) throws -> Data

    // Hash operations
    public let hash: @Sendable (Data) -> Data
    public let hmacHash: @Sendable (Data, Data) -> Data

    // Optional native HKDF override (avoids intermediate Data allocations)
    internal let hkdfOverride: (@Sendable (Data, Data, Int) -> [Data])?

    /// HKDF derived from hmacHash, or native C implementation when available.
    public func hkdf(chainingKey: Data, inputKeyMaterial: Data, numOutputs: Int) -> [Data] {
        if let native = hkdfOverride {
            return native(chainingKey, inputKeyMaterial, numOutputs)
        }
        let tempKey = hmacHash(chainingKey, inputKeyMaterial)
        let output1 = hmacHash(tempKey, hkdfCounter01)
        let output2 = hmacHash(tempKey, output1 + hkdfCounter02)
        if numOutputs == 2 { return [output1, output2] }
        let output3 = hmacHash(tempKey, output2 + hkdfCounter03)
        return [output1, output2, output3]
    }

    /// Construct full protocol name: "Noise_{pattern}_{dh}_{cipher}_{hash}"
    public func protocolName(pattern: String) -> String {
        "Noise_\(pattern)_\(dhName)_\(cipherName)_\(hashName)"
    }
}

// MARK: - HKDF counter constants (avoid per-call allocation)

private let hkdfCounter01 = Data([0x01])
private let hkdfCounter02 = Data([0x02])
private let hkdfCounter03 = Data([0x03])

// MARK: - DH helpers (shared across all suites)

private let _25519_generate: @Sendable () -> NoiseKeyPair = { NoiseKeyPair() }
private let _25519_dh: @Sendable (NoiseKeyPair, Data) throws -> Data = { kp, pub in
    try kp.dh(remotePublicKey: pub)
}
private let _25519_fromPrivate: @Sendable (Data) throws -> NoiseKeyPair = { data in
    try NoiseKeyPair(privateKeyData: data)
}

// MARK: - Cipher helpers

private let _chacha_encrypt: @Sendable (Data, UInt64, Data, Data) throws -> Data = { k, n, ad, pt in
    try NoiseCipher.encrypt(k: k, n: n, ad: ad, plaintext: pt)
}
private let _chacha_decrypt: @Sendable (Data, UInt64, Data, Data) throws -> Data = { k, n, ad, ct in
    try NoiseCipher.decrypt(k: k, n: n, ad: ad, ciphertext: ct)
}
private let _aesgcm_encrypt: @Sendable (Data, UInt64, Data, Data) throws -> Data = { k, n, ad, pt in
    try NoiseCipherAESGCM.encrypt(k: k, n: n, ad: ad, plaintext: pt)
}
private let _aesgcm_decrypt: @Sendable (Data, UInt64, Data, Data) throws -> Data = { k, n, ad, ct in
    try NoiseCipherAESGCM.decrypt(k: k, n: n, ad: ad, ciphertext: ct)
}

// MARK: - Hash helpers

private let _sha256_hash: @Sendable (Data) -> Data = { NoiseHash.hash($0) }
private let _sha256_hmac: @Sendable (Data, Data) -> Data = { NoiseHash.hmacHash(key: $0, data: $1) }
private let _sha512_hash: @Sendable (Data) -> Data = { NoiseHashSHA512.hash($0) }
private let _sha512_hmac: @Sendable (Data, Data) -> Data = { NoiseHashSHA512.hmacHash(key: $0, data: $1) }

private let _blake2s_hash: @Sendable (Data) -> Data = { input in
    var output = Data(count: 32)
    _ = input.withUnsafeBytes { inPtr in
        output.withUnsafeMutableBytes { outPtr in
            cblake2s(outPtr.baseAddress!, 32,
                     inPtr.baseAddress, inPtr.count)
        }
    }
    return output
}
private let _blake2s_hmac: @Sendable (Data, Data) -> Data = { key, data in
    var output = Data(count: 32)
    _ = key.withUnsafeBytes { keyPtr in
        data.withUnsafeBytes { dataPtr in
            output.withUnsafeMutableBytes { outPtr in
                cblake2s_hmac(outPtr.baseAddress!,
                              keyPtr.baseAddress, keyPtr.count,
                              dataPtr.baseAddress, dataPtr.count)
            }
        }
    }
    return output
}
private let _blake2b_hash: @Sendable (Data) -> Data = { input in
    var output = Data(count: 64)
    _ = input.withUnsafeBytes { inPtr in
        output.withUnsafeMutableBytes { outPtr in
            cblake2b(outPtr.baseAddress!, 64,
                     inPtr.baseAddress, inPtr.count)
        }
    }
    return output
}
private let _blake2b_hmac: @Sendable (Data, Data) -> Data = { key, data in
    var output = Data(count: 64)
    _ = key.withUnsafeBytes { keyPtr in
        data.withUnsafeBytes { dataPtr in
            output.withUnsafeMutableBytes { outPtr in
                cblake2b_hmac(outPtr.baseAddress!,
                              keyPtr.baseAddress, keyPtr.count,
                              dataPtr.baseAddress, dataPtr.count)
            }
        }
    }
    return output
}

// MARK: - Native HKDF overrides for BLAKE2 (avoids intermediate Data allocations)

private let _blake2s_hkdf: @Sendable (Data, Data, Int) -> [Data] = { ck, ikm, numOutputs in
    var out1 = Data(count: 32)
    var out2 = Data(count: 32)
    if numOutputs == 2 {
        _ = ck.withUnsafeBytes { ckPtr in
            ikm.withUnsafeBytes { ikmPtr in
                out1.withUnsafeMutableBytes { o1 in
                    out2.withUnsafeMutableBytes { o2 in
                        cblake2s_hkdf2(ckPtr.baseAddress, ckPtr.count,
                                       ikmPtr.baseAddress, ikmPtr.count,
                                       o1.baseAddress!, o2.baseAddress!)
                    }
                }
            }
        }
        return [out1, out2]
    }
    var out3 = Data(count: 32)
    _ = ck.withUnsafeBytes { ckPtr in
        ikm.withUnsafeBytes { ikmPtr in
            out1.withUnsafeMutableBytes { o1 in
                out2.withUnsafeMutableBytes { o2 in
                    out3.withUnsafeMutableBytes { o3 in
                        cblake2s_hkdf3(ckPtr.baseAddress, ckPtr.count,
                                       ikmPtr.baseAddress, ikmPtr.count,
                                       o1.baseAddress!, o2.baseAddress!, o3.baseAddress!)
                    }
                }
            }
        }
    }
    return [out1, out2, out3]
}

private let _blake2b_hkdf: @Sendable (Data, Data, Int) -> [Data] = { ck, ikm, numOutputs in
    var out1 = Data(count: 64)
    var out2 = Data(count: 64)
    if numOutputs == 2 {
        _ = ck.withUnsafeBytes { ckPtr in
            ikm.withUnsafeBytes { ikmPtr in
                out1.withUnsafeMutableBytes { o1 in
                    out2.withUnsafeMutableBytes { o2 in
                        cblake2b_hkdf2(ckPtr.baseAddress, ckPtr.count,
                                       ikmPtr.baseAddress, ikmPtr.count,
                                       o1.baseAddress!, o2.baseAddress!)
                    }
                }
            }
        }
        return [out1, out2]
    }
    var out3 = Data(count: 64)
    _ = ck.withUnsafeBytes { ckPtr in
        ikm.withUnsafeBytes { ikmPtr in
            out1.withUnsafeMutableBytes { o1 in
                out2.withUnsafeMutableBytes { o2 in
                    out3.withUnsafeMutableBytes { o3 in
                        cblake2b_hkdf3(ckPtr.baseAddress, ckPtr.count,
                                       ikmPtr.baseAddress, ikmPtr.count,
                                       o1.baseAddress!, o2.baseAddress!, o3.baseAddress!)
                    }
                }
            }
        }
    }
    return [out1, out2, out3]
}

// MARK: - 8 Cipher Suite constants

extension CipherSuite {
    /// Noise_XX_25519_ChaChaPoly_SHA256 — the default suite.
    public static let noise_25519_ChaChaPoly_SHA256 = CipherSuite(
        dhName: "25519", cipherName: "ChaChaPoly", hashName: "SHA256",
        dhlen: 32, hashlen: 32, blocklen: 64,
        generateKeyPair: _25519_generate, dh: _25519_dh, keyPairFromPrivate: _25519_fromPrivate,
        encrypt: _chacha_encrypt, decrypt: _chacha_decrypt,
        hash: _sha256_hash, hmacHash: _sha256_hmac,
        hkdfOverride: nil
    )

    /// Noise_XX_25519_ChaChaPoly_SHA512
    public static let noise_25519_ChaChaPoly_SHA512 = CipherSuite(
        dhName: "25519", cipherName: "ChaChaPoly", hashName: "SHA512",
        dhlen: 32, hashlen: 64, blocklen: 128,
        generateKeyPair: _25519_generate, dh: _25519_dh, keyPairFromPrivate: _25519_fromPrivate,
        encrypt: _chacha_encrypt, decrypt: _chacha_decrypt,
        hash: _sha512_hash, hmacHash: _sha512_hmac,
        hkdfOverride: nil
    )

    /// Noise_XX_25519_ChaChaPoly_BLAKE2s
    public static let noise_25519_ChaChaPoly_BLAKE2s = CipherSuite(
        dhName: "25519", cipherName: "ChaChaPoly", hashName: "BLAKE2s",
        dhlen: 32, hashlen: 32, blocklen: 64,
        generateKeyPair: _25519_generate, dh: _25519_dh, keyPairFromPrivate: _25519_fromPrivate,
        encrypt: _chacha_encrypt, decrypt: _chacha_decrypt,
        hash: _blake2s_hash, hmacHash: _blake2s_hmac,
        hkdfOverride: _blake2s_hkdf
    )

    /// Noise_XX_25519_ChaChaPoly_BLAKE2b
    public static let noise_25519_ChaChaPoly_BLAKE2b = CipherSuite(
        dhName: "25519", cipherName: "ChaChaPoly", hashName: "BLAKE2b",
        dhlen: 32, hashlen: 64, blocklen: 128,
        generateKeyPair: _25519_generate, dh: _25519_dh, keyPairFromPrivate: _25519_fromPrivate,
        encrypt: _chacha_encrypt, decrypt: _chacha_decrypt,
        hash: _blake2b_hash, hmacHash: _blake2b_hmac,
        hkdfOverride: _blake2b_hkdf
    )

    /// Noise_XX_25519_AESGCM_SHA256
    public static let noise_25519_AESGCM_SHA256 = CipherSuite(
        dhName: "25519", cipherName: "AESGCM", hashName: "SHA256",
        dhlen: 32, hashlen: 32, blocklen: 64,
        generateKeyPair: _25519_generate, dh: _25519_dh, keyPairFromPrivate: _25519_fromPrivate,
        encrypt: _aesgcm_encrypt, decrypt: _aesgcm_decrypt,
        hash: _sha256_hash, hmacHash: _sha256_hmac,
        hkdfOverride: nil
    )

    /// Noise_XX_25519_AESGCM_SHA512
    public static let noise_25519_AESGCM_SHA512 = CipherSuite(
        dhName: "25519", cipherName: "AESGCM", hashName: "SHA512",
        dhlen: 32, hashlen: 64, blocklen: 128,
        generateKeyPair: _25519_generate, dh: _25519_dh, keyPairFromPrivate: _25519_fromPrivate,
        encrypt: _aesgcm_encrypt, decrypt: _aesgcm_decrypt,
        hash: _sha512_hash, hmacHash: _sha512_hmac,
        hkdfOverride: nil
    )

    /// Noise_XX_25519_AESGCM_BLAKE2s
    public static let noise_25519_AESGCM_BLAKE2s = CipherSuite(
        dhName: "25519", cipherName: "AESGCM", hashName: "BLAKE2s",
        dhlen: 32, hashlen: 32, blocklen: 64,
        generateKeyPair: _25519_generate, dh: _25519_dh, keyPairFromPrivate: _25519_fromPrivate,
        encrypt: _aesgcm_encrypt, decrypt: _aesgcm_decrypt,
        hash: _blake2s_hash, hmacHash: _blake2s_hmac,
        hkdfOverride: _blake2s_hkdf
    )

    /// Noise_XX_25519_AESGCM_BLAKE2b
    public static let noise_25519_AESGCM_BLAKE2b = CipherSuite(
        dhName: "25519", cipherName: "AESGCM", hashName: "BLAKE2b",
        dhlen: 32, hashlen: 64, blocklen: 128,
        generateKeyPair: _25519_generate, dh: _25519_dh, keyPairFromPrivate: _25519_fromPrivate,
        encrypt: _aesgcm_encrypt, decrypt: _aesgcm_decrypt,
        hash: _blake2b_hash, hmacHash: _blake2b_hmac,
        hkdfOverride: _blake2b_hkdf
    )
}
